### PR TITLE
Validation with Scalaz

### PR DESCRIPTION
**Pull request by juanpedromoreno**
This PR is a join co-effort of @javipacheco and @juanpedromoreno. It brings the ability to validate forms with *scalaz*. The developed validation framework has been applied to a dummy form.

Ticket: http://47degrees.myjetbrains.com/youtrack/issue/SA-19

Please,
@javipacheco could you take a look to the UI? 
@franciscodr @fedefernandez could you review the code?

Thank you!

### CODE REVIEW

File: src/main/AndroidManifest.xml [hunk 1]
```diff
@@ -29,10 +29,10 @@
         <activity
                 android:label="@string/app_name"
                 android:name=".ui.main.MainActivity">
-            <intent-filter>
-                <action android:name="android.intent.action.MAIN"/>
-                <category android:name="android.intent.category.LAUNCHER"/>
-            </intent-filter>
+            <!--<intent-filter>-->
```
Comment (fedefernandez) [inline src/main/AndroidManifest.xml]: Maybe you forgot to restore it after the tests?

File: src/main/scala/com/fortysevendeg/scala/android/modules/validators/DummyFormValidation.scala [hunk 2]
```diff
@@ -0,0 +1,28 @@
...
+
+  val validationDescriptionAction = "saving form"
+
+  def isValidNameField: Validator[FormModel] =
+    input => hasRequiredField("name", input.name.getStringValue, validationDescriptionAction)(input)
+
+  def isValidEmailField: Validator[FormModel] =
+    input =>
+      for {
+        isFilled <- hasRequiredField("email", input.email.getStringValue, validationDescriptionAction)(input)
```
Comment (fedefernandez) [inline src/main/scala/com/fortysevendeg/scala/android/modules/validators/DummyFormValidation.scala]: What do you think about to have this strings (`email`, `age` and `name`) in `val` fields?

**Comment by fedefernandez** [comment]
I've learned and enjoyed with this code. Thanks @juanpedromoreno and @javipacheco 

Minor comments and LGTM

File: src/main/AndroidManifest.xml (hunk 1 above)
Comment (juanpedromoreno) [inline src/main/AndroidManifest.xml]: Yes, you're right... Fixing...

File: src/main/scala/com/fortysevendeg/scala/android/modules/validators/DummyFormValidation.scala (hunk 2 above)
Comment (juanpedromoreno) [inline src/main/scala/com/fortysevendeg/scala/android/modules/validators/DummyFormValidation.scala]: Good catch. I agree with you, I don't like hardcoded strings in the code. Let me think about it.

File: src/test/scala/com/fortysevendeg/scala/android/unit/models/FormValidationSpec.scala [hunk 3]
```diff
@@ -0,0 +1,135 @@
...
+
+      val valueToTest = "invalid.com"
+      val editableField = mockEditText(valueToTest)
+
+      val form = FormModel(None, Some(editableField), None)
+
+      val result = isValidEmailField(form)
+
+      result.isSuccess === false
+      result.toEither.left.get.details === s"$valueToTest is not a valid email address."
```
Comment (raulraja) [inline src/test/scala/com/fortysevendeg/scala/android/unit/models/FormValidationSpec.scala]: Codacy Issue found: [Usage of get on optional type.](https://www.codacy.com/app/47Degrees/scala-android/file/661377200/issues/source?bid=2339068&fileBranchId=2339108#l67)

File: src/test/scala/com/fortysevendeg/scala/android/unit/models/FormValidationSpec.scala [hunk 4]
```diff
@@ -0,0 +1,135 @@
...
+
+      val valueToTest = ""
+      val editableField = mockEditText(valueToTest)
+
+      val form = FormModel(Some(editableField), None, None)
+
+      val result = isValidNameField(form)
+
+      result.isSuccess === false
+      result.toEither.left.get.details === s"name can not be empty."
```
Comment (raulraja) [inline src/test/scala/com/fortysevendeg/scala/android/unit/models/FormValidationSpec.scala]: Codacy Issue found: [Usage of get on optional type.](https://www.codacy.com/app/47Degrees/scala-android/file/661377200/issues/source?bid=2339068&fileBranchId=2339108#l26)

File: src/test/scala/com/fortysevendeg/scala/android/unit/models/FormValidationSpec.scala [hunk 5]
```diff
@@ -0,0 +1,135 @@
...
+
+      val valueToTest = "AA"
+      val editableField = mockEditText(valueToTest)
+
+      val form = FormModel(None, None, Some(editableField))
+
+      val result = isValidAgeField(form)
+
+      result.isSuccess === false
+      result.toEither.left.get.details === s"$valueToTest is an invalid number."
```
Comment (raulraja) [inline src/test/scala/com/fortysevendeg/scala/android/unit/models/FormValidationSpec.scala]: Codacy Issue found: [Usage of get on optional type.](https://www.codacy.com/app/47Degrees/scala-android/file/661377200/issues/source?bid=2339068&fileBranchId=2339108#l108)

File: src/main/scala/com/fortysevendeg/scala/android/modules/utils/validation.scala [hunk 6]
```diff
@@ -0,0 +1,45 @@
...
+
+  def WithValidation[T, ValT <: T, A, S]
+  (validators: ((ValT) => Validation[ValidationMessage, T])*)
+      (happyPath: => S)
+      (implicit input: ValT, m: Manifest[T]): S = {
+
+    validators.toList.map(p => p(input).toValidationNel).sequenceU match {
+      case Failure(NonEmptyList(head: ValidationMessage)) => throw ValidationException(head)
+      case Failure(list: NonEmptyList[_]) =>
+        throw ValidationException(list.head.asInstanceOf[ValidationMessage], list.tail.asInstanceOf[List[ValidationMessage]])
```
Comment (raulraja) [inline src/main/scala/com/fortysevendeg/scala/android/modules/utils/validation.scala]: Codacy Issue found: [Usage of head/last on collections.](https://www.codacy.com/app/47Degrees/scala-android/file/661377143/issues/source?bid=2339068&fileBranchId=2339108#l30)

File: src/test/scala/com/fortysevendeg/scala/android/unit/models/FormValidationSpec.scala [hunk 7]
```diff
@@ -0,0 +1,135 @@
...
+
+      val valueToTest = ""
+      val editableField = mockEditText(valueToTest)
+
+      val form = FormModel(None, Some(editableField), None)
+
+      val result = isValidEmailField(form)
+
+      result.isSuccess === false
+      result.toEither.left.get.details === s"email can not be empty."
```
Comment (raulraja) [inline src/test/scala/com/fortysevendeg/scala/android/unit/models/FormValidationSpec.scala]: Codacy Issue found: [Usage of get on optional type.](https://www.codacy.com/app/47Degrees/scala-android/file/661377200/issues/source?bid=2339068&fileBranchId=2339108#l54)

File: src/test/scala/com/fortysevendeg/scala/android/unit/models/FormValidationSpec.scala [hunk 8]
```diff
@@ -0,0 +1,135 @@
...
+
+      val valueToTest = "4.4"
+      val editableField = mockEditText(valueToTest)
+
+      val form = FormModel(None, None, Some(editableField))
+
+      val result = isValidAgeField(form)
+
+      result.isSuccess === false
+      result.toEither.left.get.details === s"$valueToTest is an invalid number."
```
Comment (raulraja) [inline src/test/scala/com/fortysevendeg/scala/android/unit/models/FormValidationSpec.scala]: Codacy Issue found: [Usage of get on optional type.](https://www.codacy.com/app/47Degrees/scala-android/file/661377200/issues/source?bid=2339068&fileBranchId=2339108#l121)

File: src/test/scala/com/fortysevendeg/scala/android/unit/models/FormValidationSpec.scala [hunk 9]
```diff
@@ -0,0 +1,135 @@
...
+
+      val valueToTest = ""
+      val editableField = mockEditText(valueToTest)
+
+      val form = FormModel(None, None, Some(editableField))
+
+      val result = isValidAgeField(form)
+
+      result.isSuccess === false
+      result.toEither.left.get.details === s"age can not be empty."
```
Comment (raulraja) [inline src/test/scala/com/fortysevendeg/scala/android/unit/models/FormValidationSpec.scala]: Codacy Issue found: [Usage of get on optional type.](https://www.codacy.com/app/47Degrees/scala-android/file/661377200/issues/source?bid=2339068&fileBranchId=2339108#l95)

File: src/main/scala/com/fortysevendeg/scala/android/ui/validateforms/Layout.scala [hunk 10]
```diff
@@ -0,0 +1,70 @@
...
+trait Layout
+    extends ToolbarLayout
+    with Styles
+    with DummyFormValidation {
+
+  var name = slot[EditText]
+
+  var age = slot[EditText]
+
+  var email = slot[EditText]
```
Comment (raulraja) [inline src/main/scala/com/fortysevendeg/scala/android/ui/validateforms/Layout.scala]: Codacy Issue found: [Avoid mutable fields](https://www.codacy.com/app/47Degrees/scala-android/file/661377191/issues/source?bid=2339068&fileBranchId=2339108#l24)

File: src/main/scala/com/fortysevendeg/scala/android/ui/validateforms/Layout.scala [hunk 11]
```diff
@@ -0,0 +1,70 @@
...
+import macroid.{ActivityContext, AppContext}
+
+import scala.util.{Failure, Success, Try}
+
+trait Layout
+    extends ToolbarLayout
+    with Styles
+    with DummyFormValidation {
+
+  var name = slot[EditText]
```
Comment (raulraja) [inline src/main/scala/com/fortysevendeg/scala/android/ui/validateforms/Layout.scala]: Codacy Issue found: [Avoid mutable fields](https://www.codacy.com/app/47Degrees/scala-android/file/661377191/issues/source?bid=2339068&fileBranchId=2339108#l20)

File: src/main/scala/com/fortysevendeg/scala/android/ui/validateforms/Layout.scala [hunk 12]
```diff
@@ -0,0 +1,70 @@
...
+
+  var age = slot[EditText]
```
Comment (raulraja) [inline src/main/scala/com/fortysevendeg/scala/android/ui/validateforms/Layout.scala]: Codacy Issue found: [Avoid mutable fields](https://www.codacy.com/app/47Degrees/scala-android/file/661377191/issues/source?bid=2339068&fileBranchId=2339108#l22)

File: src/test/scala/com/fortysevendeg/scala/android/unit/models/FormValidationSpec.scala (hunk 3 above)
Comment (raulraja) [inline src/test/scala/com/fortysevendeg/scala/android/unit/models/FormValidationSpec.scala]: Codacy Issue found: [Usage of get on optional type.](https://www.codacy.com/app/47Degrees/scala-android/file/661377200/issues/source?bid=2339068&fileBranchId=2339108#l67)

File: src/test/scala/com/fortysevendeg/scala/android/unit/models/FormValidationSpec.scala (hunk 4 above)
Comment (raulraja) [inline src/test/scala/com/fortysevendeg/scala/android/unit/models/FormValidationSpec.scala]: Codacy Issue found: [Usage of get on optional type.](https://www.codacy.com/app/47Degrees/scala-android/file/661377200/issues/source?bid=2339068&fileBranchId=2339108#l26)

File: src/test/scala/com/fortysevendeg/scala/android/unit/models/FormValidationSpec.scala (hunk 5 above)
Comment (raulraja) [inline src/test/scala/com/fortysevendeg/scala/android/unit/models/FormValidationSpec.scala]: Codacy Issue found: [Usage of get on optional type.](https://www.codacy.com/app/47Degrees/scala-android/file/661377200/issues/source?bid=2339068&fileBranchId=2339108#l108)

File: src/main/scala/com/fortysevendeg/scala/android/modules/utils/validation.scala (hunk 6 above)
Comment (raulraja) [inline src/main/scala/com/fortysevendeg/scala/android/modules/utils/validation.scala]: Codacy Issue found: [Usage of head/last on collections.](https://www.codacy.com/app/47Degrees/scala-android/file/661377143/issues/source?bid=2339068&fileBranchId=2339108#l30)

File: src/test/scala/com/fortysevendeg/scala/android/unit/models/FormValidationSpec.scala (hunk 7 above)
Comment (raulraja) [inline src/test/scala/com/fortysevendeg/scala/android/unit/models/FormValidationSpec.scala]: Codacy Issue found: [Usage of get on optional type.](https://www.codacy.com/app/47Degrees/scala-android/file/661377200/issues/source?bid=2339068&fileBranchId=2339108#l54)

File: src/test/scala/com/fortysevendeg/scala/android/unit/models/FormValidationSpec.scala (hunk 8 above)
Comment (raulraja) [inline src/test/scala/com/fortysevendeg/scala/android/unit/models/FormValidationSpec.scala]: Codacy Issue found: [Usage of get on optional type.](https://www.codacy.com/app/47Degrees/scala-android/file/661377200/issues/source?bid=2339068&fileBranchId=2339108#l121)

File: src/test/scala/com/fortysevendeg/scala/android/unit/models/FormValidationSpec.scala (hunk 9 above)
Comment (raulraja) [inline src/test/scala/com/fortysevendeg/scala/android/unit/models/FormValidationSpec.scala]: Codacy Issue found: [Usage of get on optional type.](https://www.codacy.com/app/47Degrees/scala-android/file/661377200/issues/source?bid=2339068&fileBranchId=2339108#l95)

File: src/main/scala/com/fortysevendeg/scala/android/ui/validateforms/Layout.scala (hunk 10 above)
Comment (raulraja) [inline src/main/scala/com/fortysevendeg/scala/android/ui/validateforms/Layout.scala]: Codacy Issue found: [Avoid mutable fields](https://www.codacy.com/app/47Degrees/scala-android/file/661377191/issues/source?bid=2339068&fileBranchId=2339108#l24)

File: src/main/scala/com/fortysevendeg/scala/android/ui/validateforms/Layout.scala (hunk 11 above)
Comment (raulraja) [inline src/main/scala/com/fortysevendeg/scala/android/ui/validateforms/Layout.scala]: Codacy Issue found: [Avoid mutable fields](https://www.codacy.com/app/47Degrees/scala-android/file/661377191/issues/source?bid=2339068&fileBranchId=2339108#l20)

File: src/main/scala/com/fortysevendeg/scala/android/ui/validateforms/Layout.scala (hunk 12 above)
Comment (raulraja) [inline src/main/scala/com/fortysevendeg/scala/android/ui/validateforms/Layout.scala]: Codacy Issue found: [Avoid mutable fields](https://www.codacy.com/app/47Degrees/scala-android/file/661377191/issues/source?bid=2339068&fileBranchId=2339108#l22)

File: src/test/scala/com/fortysevendeg/scala/android/unit/models/FormValidationSpec.scala (hunk 3 above)
Comment (raulraja) [inline src/test/scala/com/fortysevendeg/scala/android/unit/models/FormValidationSpec.scala]: Codacy Issue found: [Usage of get on optional type.](https://www.codacy.com/app/47Degrees/scala-android/file/661377200/issues/source?bid=2339068&fileBranchId=2339108#l67)

File: src/test/scala/com/fortysevendeg/scala/android/unit/models/FormValidationSpec.scala (hunk 4 above)
Comment (raulraja) [inline src/test/scala/com/fortysevendeg/scala/android/unit/models/FormValidationSpec.scala]: Codacy Issue found: [Usage of get on optional type.](https://www.codacy.com/app/47Degrees/scala-android/file/661377200/issues/source?bid=2339068&fileBranchId=2339108#l26)

File: src/test/scala/com/fortysevendeg/scala/android/unit/models/FormValidationSpec.scala (hunk 5 above)
Comment (raulraja) [inline src/test/scala/com/fortysevendeg/scala/android/unit/models/FormValidationSpec.scala]: Codacy Issue found: [Usage of get on optional type.](https://www.codacy.com/app/47Degrees/scala-android/file/661377200/issues/source?bid=2339068&fileBranchId=2339108#l108)

File: src/main/scala/com/fortysevendeg/scala/android/modules/utils/validation.scala (hunk 6 above)
Comment (raulraja) [inline src/main/scala/com/fortysevendeg/scala/android/modules/utils/validation.scala]: Codacy Issue found: [Usage of head/last on collections.](https://www.codacy.com/app/47Degrees/scala-android/file/661377143/issues/source?bid=2339068&fileBranchId=2339108#l30)

File: src/test/scala/com/fortysevendeg/scala/android/unit/models/FormValidationSpec.scala (hunk 7 above)
Comment (raulraja) [inline src/test/scala/com/fortysevendeg/scala/android/unit/models/FormValidationSpec.scala]: Codacy Issue found: [Usage of get on optional type.](https://www.codacy.com/app/47Degrees/scala-android/file/661377200/issues/source?bid=2339068&fileBranchId=2339108#l54)

File: src/test/scala/com/fortysevendeg/scala/android/unit/models/FormValidationSpec.scala (hunk 8 above)
Comment (raulraja) [inline src/test/scala/com/fortysevendeg/scala/android/unit/models/FormValidationSpec.scala]: Codacy Issue found: [Usage of get on optional type.](https://www.codacy.com/app/47Degrees/scala-android/file/661377200/issues/source?bid=2339068&fileBranchId=2339108#l121)

File: src/test/scala/com/fortysevendeg/scala/android/unit/models/FormValidationSpec.scala (hunk 9 above)
Comment (raulraja) [inline src/test/scala/com/fortysevendeg/scala/android/unit/models/FormValidationSpec.scala]: Codacy Issue found: [Usage of get on optional type.](https://www.codacy.com/app/47Degrees/scala-android/file/661377200/issues/source?bid=2339068&fileBranchId=2339108#l95)

File: src/main/scala/com/fortysevendeg/scala/android/ui/validateforms/Layout.scala (hunk 10 above)
Comment (raulraja) [inline src/main/scala/com/fortysevendeg/scala/android/ui/validateforms/Layout.scala]: Codacy Issue found: [Avoid mutable fields](https://www.codacy.com/app/47Degrees/scala-android/file/661377191/issues/source?bid=2339068&fileBranchId=2339108#l24)

File: src/main/scala/com/fortysevendeg/scala/android/ui/validateforms/Layout.scala (hunk 11 above)
Comment (raulraja) [inline src/main/scala/com/fortysevendeg/scala/android/ui/validateforms/Layout.scala]: Codacy Issue found: [Avoid mutable fields](https://www.codacy.com/app/47Degrees/scala-android/file/661377191/issues/source?bid=2339068&fileBranchId=2339108#l20)

File: src/main/scala/com/fortysevendeg/scala/android/ui/validateforms/Layout.scala (hunk 12 above)
Comment (raulraja) [inline src/main/scala/com/fortysevendeg/scala/android/ui/validateforms/Layout.scala]: Codacy Issue found: [Avoid mutable fields](https://www.codacy.com/app/47Degrees/scala-android/file/661377191/issues/source?bid=2339068&fileBranchId=2339108#l22)

**Comment by javipacheco** [comment]
I have merged with master this PR and I use now EditText's errors for show the problems in the form

@raulraja @juanpedromoreno can you please review?

@fedefernandez @franciscodr thoughts?

**Comment by juanpedromoreno** [comment]
Minor comments added but also LGTM ;)

@raulraja would suggest that this validation code can be improved, and I agree with him. But maybe it makes more sense if we address it in a separate PR.

File: src/main/scala/com/fortysevendeg/scala/android/modules/utils/validation.scala [hunk 13]
```diff
@@ -0,0 +1,40 @@
...
+      errorTail: Seq[ValidationMessage] = Seq.empty)
+      extends RuntimeException(s"${errorHead.message} - ${errorHead.details}")
+
+  def WithValidation[T, ValT <: T, A, S]
+  (validators: ((ValT) => Validation[ValidationMessage, T])*)
+      (happyPath: => S)
+      (implicit input: ValT, m: Manifest[T], contextWrapper: ContextWrapper): S = {
+
+    validators.toList.map(p => p(input).toValidationNel).sequenceU match {
+      case Failure(NonEmptyList(head: ValidationMessage)) => throw ValidationException(head)
```
Comment (raulraja) [inline src/main/scala/com/fortysevendeg/scala/android/modules/utils/validation.scala]: I don't see the point of pattern matching on types like this if we are already doing a constructor match. Expected something like `Failure(NonEmptyList("error1", "error2")) match { case Failure(NonEmptyList(err, _*)) => err  }`

File: src/main/scala/com/fortysevendeg/scala/android/modules/utils/validation.scala [hunk 14]
```diff
@@ -0,0 +1,40 @@
...
+      case Failure(list: NonEmptyList[_]) =>
```
Comment (raulraja) [inline src/main/scala/com/fortysevendeg/scala/android/modules/utils/validation.scala]: No need for this. See comment above

File: src/main/scala/com/fortysevendeg/scala/android/modules/utils/validation.scala [hunk 15]
```diff
@@ -0,0 +1,40 @@
...
+import macroid.ContextWrapper
+import com.fortysevendeg.macroid.extras.ResourcesExtras._
+
+import scalaz._
+import scalaz.std.list._
+import scalaz.syntax.traverse._
+
+object validation {
+
+  type Validator[T] = (T) => Validation[ValidationMessage, T]
```
Comment (raulraja) [inline src/main/scala/com/fortysevendeg/scala/android/modules/utils/validation.scala]: We should not be using type aliases or wrappers to demonstrate this functionality as it hides the true power behind `ValidationNel`

File: src/main/scala/com/fortysevendeg/scala/android/modules/validators/DummyFormValidation.scala [hunk 16]
```diff
@@ -0,0 +1,34 @@
...
+
+  val ageField = "age"
+
+  def isValidNameField: Validator[FormModel] =
+    input => hasRequiredField(input.name, nameField, input.name.getStringValue, validationDescriptionAction)(input)
+
+  def isValidEmailField: Validator[FormModel] =
+    input =>
+      for {
+        isFilled <- hasRequiredField(input.email, emailField, input.email.getStringValue, validationDescriptionAction)(input)
```
Comment (raulraja) [inline src/main/scala/com/fortysevendeg/scala/android/modules/validators/DummyFormValidation.scala]: We should perform validation via Applicative and reducing folding over to show all errors.

**Pull request closed without merging**